File: esp32/components/graphics_lib/graphics/ui/checkbox/checkbox.cpp
#include "checkbox.h"

/*
std::string UICheckBox::render()
{
    return (selected ? ">" : "") + label + ": " + (checked ? "ON" : "OFF");
}

bool UICheckBox::on_event(const graphics_event_t event)
{
    if (event == EVENT_FOCUS)
    {
        checked = !checked;
        onChange(checked);
        return false;
    }

    return UIElement::on_event(event);
}
*/
UICheckBox::UICheckBox(ui_checkbox_config_t config, const bool defaultValue) : config(std::move(config)), checked(defaultValue)
{
}

std::string UICheckBox::render(const bool selected)
{
    return (selected ? ">" : "") + config.label + ": " + (checked ? "ON" : "OFF");
}

bool UICheckBox::on_event(const graphics_event_t event)
{
    if (event == EVENT_CLICK && config.onChange)
    {
        checked = config.onChange(!checked);
    }
    return UIElement::on_event(event);
}

File: esp32_rust/components/main_rust/src/graphics/ui/checkbox.rs
use crate::graphics::event::GraphicsEvent;
use crate::graphics::ui::element::{UIElement, UIElementState};

pub struct UICheckBox {
    label: String,
    on_change: Box<dyn Fn(bool) -> bool>,
    checked: bool,
}

impl UICheckBox {
    pub fn new<F>(
        label: String,
        on_change: F,
        default_value: bool
    ) -> Self
    where F: Fn(bool) -> bool + 'static {
        Self {
            label,
            on_change: Box::new(on_change),
            checked: default_value
        }
    }
}

impl UIElement for UICheckBox {
    fn render(&self, state: UIElementState) -> String {
        let prefix = match state {
            UIElementState::Selected => ">",
            UIElementState::SelectorPress => "\\",
            UIElementState::None => ""
        };
        let checked = if self.checked {
            "ON"
        } else {
            "OFF"
        };
        format!("{prefix}{}: {checked}", self.label)
    }

    fn on_event(&mut self, event: GraphicsEvent) -> bool {
        match event {
            GraphicsEvent::Click => {
                self.checked = (self.on_change)(!self.checked);
            }
            _ => ()
        }
        false
    }
}